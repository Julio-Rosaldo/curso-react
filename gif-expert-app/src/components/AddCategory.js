import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';

// Componente que al agregar un actualiza el array "categories"
// del componente principal, lo que desde ahi dispara que se
// renderice una nueva instancia del componente
export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState("");

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const addCategory = (categories) => {
        let categoriesAux = [inputValue];
        Array.prototype.push.apply(categoriesAux, categories);
        setCategories(categoriesAux);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        console.log(e);
        if (inputValue.trim().length > 0) {
            //setCategories(actualCategories => [...actualCategories, inputValue]);
            setCategories((categories) => {addCategory(categories)});
            setInputValue("");
        }
    }

    return (
        <Fragment>
            <form onSubmit={handleSubmit}>
                <h2>Add Category</h2>
                <input
                    type="text"
                    value={inputValue}
                    onChange={handleInputChange}>
                </input>
                <p></p>
                <button type="submit">Agregar Categoria</button>
            </form>
        </Fragment>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}