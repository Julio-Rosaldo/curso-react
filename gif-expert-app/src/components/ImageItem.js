import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';

export const ImageItem = ({image}) => {
    return (
        <div className="image animate__animated animate__bounce animate">
            <p>{image.title}</p>
            <img src={image.url} alt={image.title}></img>
        </div>
    )
}

ImageItem.propTypes = {
    image: PropTypes.object.isRequired
}