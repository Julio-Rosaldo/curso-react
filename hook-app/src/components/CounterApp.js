import React, { useState } from 'react'
import "./Counter.css"

export const CounterApp = () => {

    const initialCounters = {
        counter1: 10,
        counter2: 20
    }

    const [counters, setCounter] = useState(initialCounters);

    console.log(counters);

    function handleCounter() {
        //Clonar el objeto usando el operador spread
        let updatedCounters = {...counters};
        updatedCounters.counter1 = (counters.counter1 + 1);
        setCounter(updatedCounters);    
    }

    return (
        <div>
            <h1>Counter 1: {counters.counter1}</h1>
            <h1>Counter 2: {counters.counter2}</h1>
            <hr></hr>
            <button className="btn btn-primary" onClick={handleCounter}>+1</button>
        </div>
    )
}
