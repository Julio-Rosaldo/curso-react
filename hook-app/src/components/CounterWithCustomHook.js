import React, { Fragment } from 'react';
import "./Counter.css";
import { UseCounter } from '../hooks/UseCounter';

export const CounterWithCustomHook = () => {

    const { state, increment, decrement, reset } = UseCounter();

    return (
        <Fragment>
            <div>Counter With Custom Hook: {state}</div>
            <hr></hr>
            <button className="btn" onClick={() => increment()}>+1</button>
            <button className="btn" onClick={() => decrement()}>-1</button>
            <button className="btn" onClick={() => reset()}>Reset</button>
        </Fragment>

    )
}
