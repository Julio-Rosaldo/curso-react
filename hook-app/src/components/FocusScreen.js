import React, { useRef } from 'react';
import "./Effects.css";

export const FocusScreen = () => {

    const ref = useRef();

    const handleClick = () => {
        document.querySelector("#name").focus();
        //ref.current.focus();
    }

  return (
    <div>
        <h1>FocusScreen</h1>
        <hr></hr>
        <input id="name" className="form-control" placeholder="Nombre" 
        ref={ref}></input>

        <button className="btn btn-outline-primary mt-5" 
        onClick={handleClick}>Focus</button>
    </div>
  )
}
