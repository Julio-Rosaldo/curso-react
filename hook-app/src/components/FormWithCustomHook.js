import React, { Fragment, useEffect } from 'react';
import "./Effects.css";
import { UseForm } from '../hooks/UseForm';

export const FormWithCustomHook = () => {

    const [formValues, handleInputChange] = UseForm({
        name: "",
        email: "",
        password: ""
    });

    const {name, email, password} = formValues;

    useEffect( () => {
        console.log("Email modificado")
    }, [email] );

    const handleSubmit = (e) => {
        e.preventDefault();

        console.log(formValues);
    }

    return (
        <form onSubmit={handleSubmit}>
            <h1>FormWithCustomHook</h1>
            <hr></hr>
            <div className='form*group'>
                <input type="text" 
                name="name" 
                className="form-control" 
                placeholder="Nombre" 
                autoComplete="off" 
                value={name} 
                onChange={handleInputChange}></input>

                <input type="text" 
                name="email" 
                className="form-control" 
                placeholder="Email" 
                autoComplete="off" 
                value={email} 
                onChange={handleInputChange}></input>

                <input type="password" 
                name="password" 
                className="form-control" 
                placeholder="*****" 
                autoComplete="off" 
                value={password} 
                onChange={handleInputChange}></input>

            </div>

            <button type="submit" className="btn btn-primary">Enviar</button>
        </form>
    )
}
