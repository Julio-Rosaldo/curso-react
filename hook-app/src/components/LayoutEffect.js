import React, { useLayoutEffect, useRef } from 'react'
import { UseCounter } from '../hooks/UseCounter';
import { UseFetch } from '../hooks/UseFetch'
import "./Layout.css";

export const LayoutEffect = () => {

    const {count, increment} = UseCounter(1);

    console.log(count);

    const url = "https://breakingbadapi.com/api/quotes/" + count.toString();

    const state = UseFetch(url);
    const {loading, error, data} = state;

    //const {author, quote} = !!data && data[0];
    let quote;
    if (data != null){
        quote = data[0];
    }
    console.log(quote);

    const pRef = useRef();

    useLayoutEffect( () => {
        console.log(pRef);
        if (!loading) {
            console.log(pRef.current.clientWidth);
        }   
    }, [quote])

  return (
    <div>
        <h1>Breaking Bad Quotes</h1>

{
    loading? (
    <div className="alert alert-info text-center">Loading...</div>
    )
    : 
    (
        <blockquote className="blockquote text-right">
        <p className="mb-0" ref={pRef}>{quote.quote}</p>
        <p></p>
        <footer className="blockquote-footer">{quote.author}</footer>
        </blockquote>
    )
}


<p>{count}</p>
        
        <button className="btn btn-primary" onClick={() => increment()}>Siguiente</button>
        
        
    </div>
  )
}
