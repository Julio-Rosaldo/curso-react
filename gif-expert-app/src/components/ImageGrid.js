import React, { Fragment, useState, useEffect } from "react";
import { listGifs } from "../helpers/ListGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { ImageItem } from "./ImageItem";
import PropTypes from 'prop-types';

// Componente que recibe un elemento del arreglo "categories"
// y lo renderiza
export const ImageGrid = ({ category }) => {

    /*
    const [images, setImages] = useState([]);
    */

    // El useEffect indica que la instruccion solo se ejecutara
    // en la primera renderizacion, el array incluye los elementos
    // los cuales si cambian de valor, se debera ejecutar de nuevo
    // la renderizacion
    /*
    useEffect(() => {
        getGifs();
    }, [category])
    */

    /*
    const getGifs = () => {
        listGifs(category).then( (gifs) => {
            setImages(gifs);
        })
    }
    */

    const {data:images, loading} = useFetchGifs(category);
    console.log(images);
    console.log(loading);

    function map() {
        // La funcion <array>.map regresa una funcion for each
        // que genera un array de los resultados
        return images.map((image) => {
            return <span key={image.id}><ImageItem image={image} /></span>
        })
    }

    return (
        <Fragment>
            <h3>{category}</h3>
            {
                loading && <p>Cargando...</p>
            }
            <div className="card-grid">
                {map()}
            </div>
        </Fragment>
    )
}

ImageGrid.propTypes = {
    category: PropTypes.string.isRequired
}