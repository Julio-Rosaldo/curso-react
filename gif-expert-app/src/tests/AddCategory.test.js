import React from "react";
import "@testing-library/jest-dom";
import { AddCategory } from "../components/AddCategory";
import { shallow } from "enzyme";

describe("Pruebas de AddCategory", () => {

    const setCategories = jest.fn();
    let wrapper;

    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    })

    test("Mostrar correctamente", () => {
        
        expect(wrapper).toMatchSnapshot();
    })

    test("Cambiar valor del textbox", () => {
        const value = "Hola mundo";

        let input = wrapper.find("input");
        input.simulate("change", {
            target: {
                value: value
            }
        });

        input = wrapper.find("input");
        
        expect(input.prop("value")).toMatch(value);
    })

    test("No enviar el formulario", () => {

        wrapper.find("form").simulate("submit", { preventDefault(){} });
        expect(setCategories).not.toHaveBeenCalled();
    })

    test("LLamar setCategories", () => {
        const value = "Naruto";

        let input = wrapper.find("input");
        input.simulate("change", {
            target: {
                value: value
            }
        });

        wrapper.find("form").simulate("submit", { preventDefault(){} });
        expect(setCategories).toHaveBeenCalledTimes(1);

        input = wrapper.find("input");
        expect(input.prop("value")).toMatch("");
    })
})