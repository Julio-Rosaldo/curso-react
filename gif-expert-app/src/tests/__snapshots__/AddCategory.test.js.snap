// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de AddCategory Mostrar correctamente 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Add Category
    </h2>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <p />
    <button
      type="submit"
    >
      Agregar Categoria
    </button>
  </form>
</Fragment>
`;
